### ISSUE TITLE
`Array<contentful.EntryFieldsTypes>` Lacking Standard Array Methods

### ISSUE THREAD

**Issue by arimgibson** [contentful/contentful.js#1946]
Duplicate (with other context) of #1913 , which was recently closed

## Expected Behavior

**Bug does not seem to exist with a text array field type**

When Contentful fetches a field that is `Media, many files`, I believe the correct way to type this field is the following - `images: contentful.EntryFieldTypes.Array<contentful.EntryFieldTypes.AssetLink>;`.  With this method, I'd expect the returned data would be recognized as TypeScript as a JavaScript array with certain properties, meaning that standard Array methods (`map`, `forEach`, `length`, etc.) should work, along with other Array operations like the spread operator.

## Actual Behavior

However, this "Array" type from Contentful does not seem to include these methods and throws TypeScript errors for array methods, even though the data is a standard JavaScript array. Only type errors are thrown, but the actual data can be manipulated properly like any other array.

```ts
// Using Astro in this example, but this should be framework-agnostic

interface Props {
  title: string;
  slug: string;
  description: string;
  images: contentful.EntryFieldTypes.Array<contentful.EntryFieldTypes.AssetLink>;
}

const { images } = Astro.props;

console.log(images[0].fields.file.details);
// TS Error: Property '0' does not exist on type 'Array<AssetLink>'.

images.map((image) => image);
// TS Error: Property 'map' does not exist on type 'Array<AssetLink>'.

// The spread operator is a nifty tool that is often used to normalize certain types to standard arrays, such as NodeLists
const imagesArr = [...images];
// TS Error: Type 'Array<AssetLink>' must have a '[Symbol.iterator]()' method that returns an iterator.
```

## Possible Solution

The fix is likely as easy as making sure that the special `EntryFieldsTypes.Array` is somehow extending from TypeScript's Array property so TypeScript knows to treat it as a regular array with certain types instead of this special type.

Temporary workaround is to use `@ts-ignore` comments to quiet the errors. 

## Steps to Reproduce

1. Use the Contentful API to `getEntries` that includes a field with `Media, many files`
2. Create and assign a type to represent this data, that uses Contentful's Array -- such as `images: contentful.EntryFieldTypes.Array<contentful.EntryFieldTypes.AssetLink>;`
3. Attempt to perform certain array operations on the field -- such as `entry.fields.images.map()`

## Context

I'm trying to map the images into as Astro component; the page renders as expected but TypeScript throws the aforementioned errors.

```ts
{
  images.map((image) => (
    <Image
      src={image.fields.file.url}
      alt={image.fields.title}
      width={image.fields.file.details.image.width}
      height={image.fields.file.details.image.height}
    />
  ))
}
```

## Environment

-   **Language Version**: v18.16.0
-   **Package Manager Version**: pnpm v8.6.2
-   **Operating System**: Fedora Linux 38
-   **Package Version**: 10.3.0
-   **TypeScript Version**: 5.1.3
-   **Which API are you using?**: Preview

**Comment by arimgibson** [comment]
Okay, typing the `images` directly as an Array of Contentful Assets instead of using the `EntryFieldTypes.AssetLink` worked for me. The `AssetLink` typing seems more intuitive though, as it aligns with the rest of my typing:

```ts
interface ContentfulPortfolioEntryStaticImagesCarousels {
  contentTypeId: "portfolioEntryStaticImagesCarousels";
  fields: {
    title: contentful.EntryFieldTypes.Symbol;
    slug: contentful.EntryFieldTypes.Symbol;
    description: contentful.EntryFieldTypes.Text;
    images: contentful.EntryFieldTypes.Array<contentful.EntryFieldTypes.AssetLink>;
    firstCarousel: contentful.EntryFieldTypes.Array<contentful.EntryFieldTypes.AssetLink>;
    secondCarousel?: contentful.EntryFieldTypes.Array<contentful.EntryFieldTypes.AssetLink>;
  };
}
```

And here's the example of typing as Assets directly:
```ts
interface ContentfulPortfolioEntryStaticImagesCarousels {
  contentTypeId: "portfolioEntryStaticImagesCarousels";
  fields: {
    title: contentful.EntryFieldTypes.Symbol;
    slug: contentful.EntryFieldTypes.Symbol;
    description: contentful.EntryFieldTypes.Text;
    images: Array<contentful.Asset>;
    firstCarousel: Array<contentful.Asset>;
    secondCarousel?: Array<contentful.Asset>;
  };
}
```

Leaving open as I still believe this might be a bug.

**Comment by arimgibson** [comment]
Mmmmm, good to know @veu! I was passing the generic in but my types were getting too complex to continue using the implicit types returned by Contentful so I had turned to using the same interface for explicit typing.

This probably isn't the most elegant way, but the best in my situation instead of type inference:  I chose to create new explicit types that would represent that data I would be using after all processing, as I was immediately mapping the Contentful entries to a different structure. Now I can use these explicit types across my codebase while also using the `as` keyword to keep things consistent, instead of switching between my own types and the types from Contentful.
